annnufan/assignment2/node.cpp

```diff
@@ -0,0 +1,96 @@
+#include <iostream>
+#include <algorithm>
+
+
+class node {
+	int key;
+public:
+	node(int x, node* l = nullptr, node* r = nullptr) : key(x), left_node(l), right_node(r){}
+	~node() {
+		delete left_node;
+		delete right_node;
+	}
+
+	void set_value(int x) {
+		key = x;
+	}
+
+	int get_value() {
+		return key;
+	}
+
+	node* left_node, *right_node, *parent_node;
+};
+
+bool find_key_and_print_way_to_root(int x, node* v) {
+	if (v == nullptr) {
+		return false;
+	}
+	if (x == v->get_value()) {
+		return true;
+	}
+	if (find_key_and_print_way_to_root(x, v->left_node) || find_key_and_print_way_to_root(x, v->right_node)) {
+		std::cout << v->get_value() << ' ';
+		return true;
+	}
+	return false;
+}
+
+void print_way_to_root_for_key(int x, node* v) {
+	if (find_key_and_print_way_to_root(x, v)) {
+		std::cout << "\n";
+	} else {
+		std::cout << "Couldn't find key\n";
+	}
+}
+
+node* find_node(int x, node* v, node* p) {

```

Comment:
You already have a function above `find_key_and_print_way_to_root`. Maybe you can extract common logic to a function and avoid code duplication? For example, both solutions can re-use the same `find_key` function.